ubr/report.py

```diff
@@ -0,0 +1,139 @@
+
+    # we want to transform the deeply nested struct above to retain only the most recent
+    # backup. the leaves are sorted lists in ascending order, least to most recent
+
+    def apply_to_backup(x):
+        return isinstance(x, list)
+
+    def most_recent_backup(lst):
+        return lst[-1]
+
+    # {project: {host: {filename: latest-item}}}
+    return visit(backup_list, apply_to_backup, most_recent_backup)
+
+
+def dtobj_from_backup(backup):
+    "given a backup struct, returns a datetime object"
+    dtstr = backup["ymd"] + backup["hms"]
+    return datetime.strptime(dtstr, "%Y%m%d%H%M%S")
+
+
+def old_backup(backup):
+    "predicate, returns true if given backup is 'old' (older than 2 days)"
+    dtobj = dtobj_from_backup(backup)
+    diff = datetime.now() - dtobj
+    threshold = 2  # days
+    return diff.days > threshold
+
+
+#
+
+
+def print_report(backup_list):
+    "given a list of backups, prints it's details"
+    result = group_by_many(backup_list, ["project", "host", "filename"])
+    for project, hosts in result.items():
+        print(project)
+        for host, files in hosts.items():
+            print(" ", host)
+            for filename, backup in files.items():
+                backup = backup[0]  # why is this a list ... ?
+                ppdt = dtobj_from_backup(backup)
+                print("   %s: %s" % (filename, ppdt))
+
+
+def check_all():
+    "check all project that backups are happening"
+    results = all_projects_latest_backups_by_host_and_filename(conf.BUCKET)
+    problems = []
+    for project, hosts in results.items():
+        for host, files in hosts.items():
+            for filename, backup in files.items():
+                old_backup(backup) and problems.append(backup)
+                # problems.append(backup)
+    problems and print_report(problems)
+    return problems
+
+
+def check(hostname, path_list=None):
+    "check self that backups are happening"
+    problems = []
+    for descriptor_path in find_descriptors(conf.DESCRIPTOR_DIR):
+        # 'lax'
+        project = pname(descriptor_path)
+        # {'postgresql-database': ['lax']}
+        descriptor = load_descriptor(descriptor_path, path_list)
+        for target, remote_path_list in descriptor.items():
+            # [('laxprod-psql.gz', 'lax/201908/20190825_prod--lax.elifesciences.org_230337-laxprod-psql.gz')
+            #  ('laxprod-archive.tar.gz', ' 'lax/201908/20190825_prod--lax.elifesciences.org_230337-laxprod-psql.gz')]
+            latest_for_target = s3.latest_backups(
+                conf.BUCKET, project, hostname, target
+            )
+            path_list = [s3_path for fname, s3_path in latest_for_target]
+            backup_list = parse_prefix_list(path_list)
+            for backup in backup_list:
+                old_backup(backup) and problems.append(backup)
+                #problems.append(backup)
+    problems and print_report(problems)
```

Comment:
I don't find easy to understand side effects like print hidden inside a logical operator, but if it works for you ok

Reply:
really? that's ok. think of it like "when problems, call function". it really is just an artefact of me switching between languages in my head. I'll change it for you